File: docs/xplat/scripts/platform_blocks.py
"""Depth-aware reading of <PlatformBlock> spans.

These nest, so pairing an opener with the first closer after it takes the inner block's closer — the
body then stops halfway and the tail looks like it belongs to no block at all.
"""
import re

_OPEN = re.compile(r'<PlatformBlock\s+for="([^"]+)">')
_CLOSE = re.compile(r'</PlatformBlock>')


def blocks_of(text):
    """(start, end, body_start, body_end, platforms, children) for every block, in order."""
    events = sorted([(m.start(), m.end(), 'open', m.group(1)) for m in _OPEN.finditer(text)] +
                    [(m.start(), m.end(), 'close', None) for m in _CLOSE.finditer(text)])
    stack, out = [], []
    for start, end, kind, platforms in events:
        if kind == 'open':
            stack.append([start, end, platforms, 0])
            continue
        if not stack:
            continue                                  # a stray closer; the page is unbalanced
        opened = stack.pop()
        out.append({'start': opened[0], 'end': end, 'body_start': opened[1], 'body_end': start,
                    'platforms': [p.strip() for p in opened[2].split(',')],
                    'children': opened[3]})
        if stack:
            stack[-1][3] += 1
    return sorted(out, key=lambda b: b['start'])


def leaf_blocks_of(text):
    """The blocks holding no other block, which is where a topic's content lives."""
    return [b for b in blocks_of(text) if b['children'] == 0]

File: docs/xplat/scripts/platform-blocks.mjs
/**
 * Depth-aware reading of <PlatformBlock> spans.
 *
 * These nest — a Web Components block inside an Angular one is common in the grid topics — so
 * pairing an opener with the first closer after it takes the *inner* block's closer. Everything
 * downstream is then working with a body that stops halfway and a tail that looks like it belongs to
 * no block at all. That mistake silently deleted three sections from a Japanese topic before this
 * existed, so anything that needs a block's extent uses this rather than a regex of its own.
 */

const OPEN = /<PlatformBlock\s+for="([^"]+)">/g;
const CLOSE = /<\/PlatformBlock>/g;

/**
 * Every block in the text, innermost first, each with:
 *   start, end        the whole element, opening tag through closing tag
 *   bodyStart, bodyEnd  what is between the tags
 *   platforms         the for= list, split and trimmed
 *   depth             1 for a top level block
 *   children          how many blocks it directly contains
 */
export function blocksOf(text) {
    const events = [];
    for (const m of text.matchAll(OPEN)) {
        events.push({ at: m.index, after: m.index + m[0].length, kind: 'open', list: m[1] });
    }
    for (const m of text.matchAll(CLOSE)) {
        events.push({ at: m.index, after: m.index + m[0].length, kind: 'close' });
    }
    events.sort((a, b) => a.at - b.at);

    const stack = [], blocks = [];
    for (const e of events) {
        if (e.kind === 'open') {
            stack.push({ start: e.at, bodyStart: e.after, list: e.list, children: 0 });
            continue;
        }
        const frame = stack.pop();
        if (!frame) continue;                       // a stray closer; the page is unbalanced
        blocks.push({
            start: frame.start, end: e.after,
            bodyStart: frame.bodyStart, bodyEnd: e.at,
            platforms: frame.list.split(',').map(s => s.trim()),
            depth: stack.length + 1,
            children: frame.children,
        });
        if (stack.length) stack[stack.length - 1].children++;
    }
    return blocks.sort((a, b) => a.start - b.start);
}

/** The blocks that contain no other block, which is where a topic's content actually lives. */
export function leafBlocksOf(text) {
    return blocksOf(text).filter(b => b.children === 0);
}

/** Whether the text's blocks are balanced, which every reader here assumes. */
export function isBalanced(text) {
    const opens = (text.match(/<PlatformBlock\s+for="[^"]+">/g) || []).length;
    const closes = (text.match(/<\/PlatformBlock>/g) || []).length;
    return opens === closes;
}
